src/test/java/com/amazon/ion/bytecode/util/ConstantPoolTest.kt

```diff
@@ -0,0 +1,480 @@
+        val smallPool = ConstantPool(2)
+
+        // Add items beyond initial capacity
+        for (i in 0..5) {
+            val index = smallPool.add("item$i")
+            Assertions.assertEquals(i, index)
+        }
+
+        Assertions.assertEquals(6, smallPool.size)
+
+        // Verify all items are accessible
+        for (i in 0..5) {
+            Assertions.assertEquals("item$i", smallPool.get(i))
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = [1, 5, 10, 100, 1000])
+    fun `pool handles various initial capacities`(initialCapacity: Int) {
+        val pool = ConstantPool(initialCapacity)
+        Assertions.assertTrue(pool.isEmpty())
+        Assertions.assertEquals(0, pool.size)
+
+        // Add one item to verify it works
+        val index = pool.add("test")
+        Assertions.assertEquals(0, index)
+        Assertions.assertEquals("test", pool.get(0))
+    }
+
+    @Test
+    fun `toArray returns defensive copy with correct size`() {
+        val constantPool = ConstantPool(10)
+        constantPool.add("item1")
+        constantPool.add("item2")
+        constantPool.add(null)
+
+        val array = constantPool.toArray()
+        Assertions.assertEquals(3, array.size)
+        Assertions.assertEquals("item1", array[0])
+        Assertions.assertEquals("item2", array[1])
+        Assertions.assertNull(array[2])
+
+        // Verify it's a defensive copy by modifying the returned array
+        array[0] = "modified"
+        Assertions.assertEquals("item1", constantPool.get(0)) // Original should be unchanged
+    }
+
+    @Test
+    fun `toArray returns empty array for empty pool`() {
+        val constantPool = ConstantPool(10)
+        val array = constantPool.toArray()
+        Assertions.assertEquals(0, array.size)
+    }
+
+    @Test
+    fun `unsafeGetArray returns backing array`() {
+        val constantPool = ConstantPool(10)
+        constantPool.add("item1")
+        constantPool.add("item2")
+
+        val array = constantPool.unsafeGetArray()
+
+        // Array should contain the items
+        Assertions.assertEquals("item1", array[0])
+        Assertions.assertEquals("item2", array[1])
+
+        // Array size should be capacity, not just the number of items
+        Assertions.assertTrue(array.size >= constantPool.size)
+    }
+
+    @Test
+    fun `toString returns correct string representation`() {
+        val constantPool = ConstantPool(10)
+        val emptyString = constantPool.toString()
+        Assertions.assertEquals("ConstantPool(data=[])", emptyString)
+
+        constantPool.add("test")
+        constantPool.add(42)
+        constantPool.add(null)
+
+        val string = constantPool.toString()
+        Assertions.assertTrue(string.startsWith("ConstantPool(data=["))
+        Assertions.assertTrue(string.contains("test,"))
+        Assertions.assertTrue(string.contains("42,"))
+        Assertions.assertTrue(string.contains("null,"))
```

Comment:
Is this the same output as if we had added `constantPool.add("null")`. (Note: this would be a nitpick at best)